### PR TITLE
adds stl, thrust, opencv interop functions

### PR DESCRIPTION
**Pull request by syurkevi**
Addresses interop issues: #1504 , #1505, #529 
Rider issue #1503 .

### CODE REVIEW

File: include/af/array.h [hunk 1]
```diff
@@ -12,10 +12,57 @@
...
+    */
+
+    /**
+       \ingroup method_mat
+       @{
+    */
+    /**
+       \brief Reduce the reference count of the \ref af_array
+    */
+    AFAPI af_err af_release_array(af_array arr);
```
Comment (umar456) [inline include/af/array.h]: why are these moved?

File: include/af/opencv2.hpp [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+/*
+ * arrayfire -> opencv
+ */
+// af::array conversion to cv::Mat
+cv::Mat toMat(const af::array& input, int type = CV_32F, bool transpose = true);
+
+
+//helper conversion function
+void copy_data(Mat& input, af::array& output, bool transpose) {
```
Comment (umar456) [inline include/af/opencv2.hpp]: I think this will throw an error if included in multiple files in the same binary

File: include/af/opencv2.hpp [hunk 3]
```diff
@@ -0,0 +1,160 @@
+#pragma once
```
Comment (umar456) [inline include/af/opencv2.hpp]: The file should be `opencv.h`

File: include/af/opencv3.hpp [hunk 4]
```diff
@@ -0,0 +1,160 @@
+#pragma once
```
Comment (umar456) [inline include/af/opencv3.hpp]: rename to `opencv3.h`

File: include/af/opencv2.hpp (hunk 2 above)
Comment (syurkevi) [inline include/af/opencv2.hpp]: Only one should be included at a time. Easy workaround would be to use:
#include "opencv2/core/version.hpp"
#if CV_MAJOR_VERSION == 2
#include opencv2.hpp
#elif CV_MAJOR_VERSION == 3
#include opencv3.hpp
#endif

File: include/af/opencv2.hpp [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+#include <vector>
+#include <opencv2/core/core.hpp>
+#include <arrayfire.h>
+
+namespace afcv
```
Comment (umar456) [inline include/af/opencv2.hpp]: i think we should have a subnamespace instead of this.
i.e.
namespace af {
namespace opencv {
}
}

File: include/af/opencv2.hpp [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+{
+
+/*
+ * opencv -> arrayfire
```
Comment (umar456) [inline include/af/opencv2.hpp]: Document

File: include/af/opencv2.hpp [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+ */
+// conversion for cv::Mat
+af::array array(const std::vector<cv::Mat>& input, bool transpose = true);
+af::array array(const cv::Mat& input, bool transpose = true);
+
+/*
+ * arrayfire -> opencv
```
Comment (umar456) [inline include/af/opencv2.hpp]: document

File: include/af/array.h (hunk 1 above)
Comment (syurkevi) [inline include/af/array.h]: They are used in the new assignment operators and need to be defined before since the new operators are templated.

File: include/af/opencv2.hpp [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+ * opencv -> arrayfire
+ */
+
+// conversion for cv::Mat
+af::array array(const cv::Mat& input, bool transpose)
+{
+    af::array output;
+    if (input.empty()) { return output; }
+    cv::Mat tmp;
+    if (input.channels() == 1)
```
Comment (umar456) [inline include/af/opencv2.hpp]: will this only work for float types?

File: include/af/opencv3.hpp [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        int* data = output.ptr<int>(0);
+        input_.as(s32).host((void*)data);
+    } else if (type == CV_64F) {
+        double* data = output.ptr<double>(0);
+        input_.as(f64).host((void*)data);
+    } else if (type == CV_8U) {
+        uchar* data = output.ptr<uchar>(0);
+        input_.as(b8).host((void*)data);
+    } else {
+        throw std::runtime_error(string("array to mat error "));
```
Comment (umar456) [inline include/af/opencv3.hpp]: Use af::exception instead

File: src/api/cpp/array.cpp [hunk 10]
```diff
@@ -1001,6 +1002,17 @@ af::dtype implicit_dtype(af::dtype scalar_type, af::dtype array_type)
...
 
+#define INSTANTIATE(T)                                              \
+    template<> AFAPI T* array::device() const                       \
+    {                                                               \
+        void *ptr = NULL;                                           \
+        AF_THROW(af_get_device_ptr(&ptr, get()));                   \
+        return (T *)ptr;                                            \
+    }
+
+INSTANTIATE(void)
```
Comment (umar456) [inline src/api/cpp/array.cpp]: seems unnecessary to make a macro for one type

**Comment by umar456** [comment]
I don't see tests for these functions

File: include/af/array.h [hunk 11]
```diff
@@ -682,6 +729,64 @@ namespace af
...
+            af_array temp = 0;
+            dim_t sz = other.size();
+            af_dtype type = (af_dtype)dtype_traits<T>::af_type;
+            af_err err = af_create_array(&temp, &other[0], 1, &sz, type);
+            if(err != AF_SUCCESS) { throw af::exception(err); }
+            this->arr = temp;
+            return *this;
+        }
+
+        template<typename T> operator std::vector<T>() {
```
Comment (umar456) [inline include/af/array.h]: I really really really don't want this function in arrayfire. It will allow the following code to work:

```
vector<float> my_funciton(vector<float>& foo) {
  for(auto &b : foo) {
    b++;
  }
  return foo;
}
int main () {
  af::array bar = randu(10, 10);
  bar = my_function(bar);
  return 0;
}
```

File: include/af/opencv2.hpp (hunk 2 above)
Comment (umar456) [inline include/af/opencv2.hpp]: I mean if you included this specific header into two or more cpp source files then you will encounter a multiple definition error because this is not a template function

File: include/af/opencv2.hpp (hunk 5 above)
Comment (pavanky) [inline include/af/opencv2.hpp]: This is not consistent with what we are doing with other interop namespaces.

File: src/api/cpp/array.cpp (hunk 10 above)
Comment (pavanky) [inline src/api/cpp/array.cpp]: We have the macro defined already some other place. Use that instead.

File: include/af/opencv2.hpp (hunk 2 above)
Comment (pavanky) [inline include/af/opencv2.hpp]: @syurkevi Umar is right, make the functions static.

File: include/af/cuda.h [hunk 12]
```diff
@@ -117,5 +122,93 @@ static inline void setNativeId(int nativeId)
...
+/**
+   Interop function that converts a thrust::device_vector to an af::array
+
+   \param[in] tvec thrust::device_vector to be converted
+   \return af::array with converted contents of the thrust vector
+
+   \ingroup cuda_mat
+ */
+template<typename T>
+static af::array array(thrust::device_vector<T> &tvec)
```
Comment (pavanky) [inline include/af/cuda.h]: Templated functions need not be static.

File: include/af/cuda.h [hunk 13]
```diff
@@ -117,5 +122,93 @@ static inline void setNativeId(int nativeId)
...
+{
+    af::array output;
+    if(tvec.empty()) return output;
+    T* ptr = thrust::raw_pointer_cast(tvec.data());
+    af::array tmp = af::array(tvec.size(), ptr, afDevice);
+    output = tmp.copy();
+    tmp.lock();
```
Comment (pavanky) [inline include/af/cuda.h]: Don't perform a copy by default. Have a flag that is set to false and copy when the user requests it.

File: include/af/cuda.h [hunk 14]
```diff
@@ -117,5 +122,93 @@ static inline void setNativeId(int nativeId)
...
+   \return af::array with converted contents of the thrust vector
+
+   \ingroup cuda_mat
+ */
+template<typename T>
+static af::array array(thrust::host_vector<T> &tvec)
+{
+    af::array output;
+    if(tvec.empty()) return output;
+    //af_dtype type = (af_dtype)dtype_traits<T>::af_type;
```
Comment (pavanky) [inline include/af/cuda.h]: Remove this.

File: include/af/cuda.h [hunk 15]
```diff
@@ -117,5 +122,93 @@ static inline void setNativeId(int nativeId)
...
+   \param[in] arr af::array to be converted to thrust::host_vector
+   \return thrust::host_vector with converted contents of af::array
+
+   \ingroup cuda_mat
+ */
+template<typename T>
+static thrust::host_vector<T> toHostVector(af::array &arr)
+{
+    af_dtype vec_type = (af_dtype)af::dtype_traits<T>::af_type;
+    if(arr.type() != vec_type) { throw af::exception("Thrust vector data type and array data type are mismatching"); }
```
Comment (pavanky) [inline include/af/cuda.h]: throw in new line.

File: include/af/cuda.h [hunk 16]
```diff
@@ -117,5 +122,93 @@ static inline void setNativeId(int nativeId)
...
+    //todo check continuity
+    thrust::host_vector<T> hvec(arr.elements());
+    T* h_ptr = thrust::raw_pointer_cast(hvec.data());
+    arr.host(h_ptr);
```
Comment (pavanky) [inline include/af/cuda.h]: `.host()` does the proper transformations to make it linear. You don't need to check for continuity.

File: include/af/cuda.h [hunk 17]
```diff
@@ -117,5 +122,93 @@ static inline void setNativeId(int nativeId)
...
+ */
+template<typename T>
+static thrust::device_vector<T> toDeviceVector(af::array &arr)
+{
+    af_dtype vec_type = (af_dtype)af::dtype_traits<T>::af_type;
+    if(arr.type() != vec_type) { throw af::exception("Thrust vector data type and array data type are mismatching"); }
+    arr.lock();
+    //todo check continuity
+    thrust::device_ptr<T> af_ptr = thrust::device_pointer_cast(arr.device<T>());
+    thrust::device_vector<T> dvec(af_ptr, af_ptr + arr.elements());
```
Comment (pavanky) [inline include/af/cuda.h]: Same with `.device`

File: include/af/cuda.h [hunk 18]
```diff
@@ -117,5 +122,93 @@ static inline void setNativeId(int nativeId)
...
+   \param[in] arr af::array to be converted to thrust::device_vector
+   \return thrust::device_vector with converted contents of af::array
+
+   \ingroup cuda_mat
+ */
+template<typename T>
+static thrust::device_vector<T> toDeviceVector(af::array &arr)
+{
+    af_dtype vec_type = (af_dtype)af::dtype_traits<T>::af_type;
+    if(arr.type() != vec_type) { throw af::exception("Thrust vector data type and array data type are mismatching"); }
```
Comment (pavanky) [inline include/af/cuda.h]: throw in new line.

File: include/af/cuda.h [hunk 19]
```diff
@@ -117,5 +122,93 @@ static inline void setNativeId(int nativeId)
...
+    arr.lock();
+    //todo check continuity
+    thrust::device_ptr<T> af_ptr = thrust::device_pointer_cast(arr.device<T>());
+    thrust::device_vector<T> dvec(af_ptr, af_ptr + arr.elements());
+    arr.unlock();
```
Comment (pavanky) [inline include/af/cuda.h]: `unlock` needs to be called when `device_vector` goes out of scope. We can do this but it would require a little bit more work.

File: include/af/array.h [hunk 20]
```diff
@@ -682,6 +729,64 @@ namespace af
            @}
         */
 
+#ifdef AF_ENABLE_STL_CAST
+        template<typename T> array& operator=(const std::vector<T> &other)
+        {
+            //TODO: Unsafe. loses data if af_weak_copy fails
+            if(this->arr != 0) {
+                af_err err = af_release_array(this->arr);
+                if(err != AF_SUCCESS) { throw af::exception(err); }
```
Comment (pavanky) [inline include/af/array.h]: Have a message in the exception.

File: include/af/array.h [hunk 21]
```diff
@@ -682,6 +729,64 @@ namespace af
...
+            }
+
+            af_array temp = 0;
+            dim_t sz = other.size();
+            af_dtype type = (af_dtype)dtype_traits<T>::af_type;
+            af_err err = af_create_array(&temp, &other[0], 1, &sz, type);
+            if(err != AF_SUCCESS) { throw af::exception(err); }
```
Comment (pavanky) [inline include/af/array.h]: Have a message in the exception.

File: include/af/opencv3.h [hunk 22]
```diff
@@ -0,0 +1,191 @@
...
+   \returns an cv::Mat with the contents of the input af::array
+*/
+static cv::Mat toMat(const af::array& input, int type = CV_32F, bool transpose = true);
+/**
+  @}
+*/
+
+
+//helper conversion function
+static void copy_data(cv::Mat& input, af::array& output, bool transpose) {
```
Comment (pavanky) [inline include/af/opencv3.h]: just `copy` is fine. Set `transpose=false` as default value.

File: include/af/opencv3.h [hunk 23]
```diff
@@ -0,0 +1,191 @@
...
+/**
+   This function will copy data from an \ref af::array to a cv::Mat
+
+   \param[in]  input The af::array from which data which will be loaded into cv::Mat
+   \param[in]  type The resulting datatype of the output cv::Mat
+   \param[in]  transpose flag specifying if the input should be transposed during copy
+
+   \returns an cv::Mat with the contents of the input af::array
+*/
+static cv::Mat toMat(const af::array& input, int type = CV_32F, bool transpose = true);
```
Comment (pavanky) [inline include/af/opencv3.h]: No reason to separate definition and declaration.

File: include/af/opencv3.h [hunk 24]
```diff
@@ -0,0 +1,191 @@
...
+  @}
+*/
+
+
+//helper conversion function
+static void copy_data(cv::Mat& input, af::array& output, bool transpose) {
+    const unsigned w = input.cols;
+    const unsigned h = input.rows;
+    const unsigned nchannels = input.channels();
+    if (nchannels > 4) { throw af::exception("mat to array error "); }
```
Comment (pavanky) [inline include/af/opencv3.h]: Explain why it is failing. The message does not say much.

File: include/af/opencv3.h [hunk 25]
```diff
@@ -0,0 +1,191 @@
...
+    if (input.channels() == 1)
+        { input.convertTo(tmp, CV_32F); }
+    else if (input.channels() == 2)
+        { input.convertTo(tmp, CV_32FC2); }
+    else if (input.channels() == 3)
+        { input.convertTo(tmp, CV_32FC3); }
+    else if (input.channels() == 4)
+        { input.convertTo(tmp, CV_32FC4); }
+
+    copy_data(tmp, output, transpose);
```
Comment (pavanky) [inline include/af/opencv3.h]: There are several things wrong here.

- copy_data is only used once in this file.
- It should also not be used by the user because it performs no checks.
- The function is not actually performing a copy. It is creating a new output array.

Just move all the relevant code here.

File: include/af/opencv3.h [hunk 26]
```diff
@@ -0,0 +1,191 @@
...
+    const int channels = input.dims(2);
+    int ndims = input.numdims();
+    af::array input_;
+    if (transpose) {
+        if(channels == 1) {
+            input_ = input;
+        } else {
+            input_ = af::constant(0, channels, input.dims(1), input.dims(0));
+            for(int ii=0; ii<channels; ++ii) {
+                input_(ii, af::span, af::span) = af::moddims(input(af::span, af::span, ii), 1, input.dims(1), input.dims(0));
```
Comment (pavanky) [inline include/af/opencv3.h]: Use multiple lines for RHS.

File: include/af/opencv3.h [hunk 27]
```diff
@@ -0,0 +1,191 @@
...
+            }
+        }
+        output = cv::Mat(input.dims(1), input.dims(0), CV_MAKETYPE(type, channels));
+    } else {
+        if(channels == 1) {
+            input_ = input.T();
+        } else {
+            input_ = af::constant(0, channels, input.dims(0), input.dims(1));
+            for(int ii=0; ii<channels; ++ii) {
+                input_(ii, af::span, af::span) = af::moddims(input(af::span, af::span, ii).T(), 1, input.dims(0), input.dims(1));
```
Comment (pavanky) [inline include/af/opencv3.h]: Use multiple lines for RHS

File: include/af/opencv3.h [hunk 28]
```diff
@@ -0,0 +1,191 @@
...
+    int ndims = input.numdims();
+    af::array input_;
+    if (transpose) {
+        if(channels == 1) {
+            input_ = input;
+        } else {
+            input_ = af::constant(0, channels, input.dims(1), input.dims(0));
+            for(int ii=0; ii<channels; ++ii) {
+                input_(ii, af::span, af::span) = af::moddims(input(af::span, af::span, ii), 1, input.dims(1), input.dims(0));
+            }
```
Comment (pavanky) [inline include/af/opencv3.h]: This is just a reorder operation. Remove for loop. `array input_ = af::reorder(input, 2, 1, 0);`

File: include/af/opencv3.h [hunk 29]
```diff
@@ -0,0 +1,191 @@
...
+        }
+        output = cv::Mat(input.dims(1), input.dims(0), CV_MAKETYPE(type, channels));
+    } else {
+        if(channels == 1) {
+            input_ = input.T();
+        } else {
+            input_ = af::constant(0, channels, input.dims(0), input.dims(1));
+            for(int ii=0; ii<channels; ++ii) {
+                input_(ii, af::span, af::span) = af::moddims(input(af::span, af::span, ii).T(), 1, input.dims(0), input.dims(1));
+            }
```
Comment (pavanky) [inline include/af/opencv3.h]: This is just a reorder operation. Remove for loop. `array input_ = af::reorder(input, 2, 0, 1);`

File: include/af/opencv.h [hunk 30]
```diff
@@ -0,0 +1,190 @@
+#pragma once
+#include <vector>
+#include <opencv2/core/core.hpp>
+#include <arrayfire.h>
+
```
Comment (pavanky) [inline include/af/opencv.h]: You include opencv2 both times.

**Comment by pavanky** [comment]
Moving to https://github.com/arrayfire/arrayfire-extras
